src/repr/src/row.rs

```diff
@@ -832,6 +860,72 @@ impl RowPacker {
         res
     }
 
+    /// Convenience function to construct an array from an iter of `Datum`s.
+    ///
+    /// Returns an error if the number of elements in `iter` does not match
+    /// the cardinality of the array as described by `dims`, or if the
+    /// number of dimensions exceeds [`MAX_ARRAY_DIMENSIONS`]. If an error
+    /// occurs, the packer's state will be unchanged.
+    pub fn push_array<'a, I, D>(
+        &mut self,
+        dims: &[ArrayDimension],
+        iter: I,
+    ) -> Result<(), InvalidArrayError>
+    where
+        I: IntoIterator<Item = D>,
+        D: Borrow<Datum<'a>>,
+    {
+        // Arrays are encoded as follows.
+        //
+        // u8      ndims
+        // usize   dim_0 lower bound
+        // usize   dim_0 length
+        // ...
+        // usize   dim_n lower bound
+        // usize   dim_n length
+        // usize   element data size in bytes
+        // u8      element data, where elements are encoded in row-major order
+
+        if dims.len() >= usize::from(MAX_ARRAY_DIMENSIONS) {
+            return Err(InvalidArrayError::TooManyDimensions(dims.len()));
+        }
+
+        let start = self.data.len();
+        self.data.push(Tag::Array as u8);
+
+        // Write dimension information.
+        self.data
+            .push(dims.len().try_into().expect("ndims verified to fit in u8"));
+        for dim in dims {
+            push_copy!(&mut self.data, dim.lower_bound, usize);
+            push_copy!(&mut self.data, dim.length, usize);
+        }
+
+        // Write elements.
+        let off = self.data.len();
+        push_copy!(&mut self.data, 0, usize); // dummy length fixed up below
```

Comment:
Two questions:

1. Why do we need this at all, since the length can be computed as the product of dim lengths?
2. Assuming we do need it, why do you need to do the "write zero and then fixup" trick?  You can just move up line 917 to before this, and write down the correct length here

Reply:
This is length in bytes, not number of elements. Like, `text[]` will be variable size in bytes even though the number of elements is fixed.

Reply:
I think this is the number of bytes the data took when encoded, not the number of datums.

Reply:
ok, that makes more sense.